packages/flutter/lib/src/rendering/object.dart

```diff
@@ -1120,6 +1401,234 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
+  /// value true to reduce the cost. If semantics are being added or
+  /// removed, more work needs to be done to update the semantics
+  /// tree. If you pass 'onlyChanges: true' but this node, which
+  /// previously had a SemanticsNode, no longer has one, or previously
+  /// did not set any semantics, but now does, or previously had a
+  /// child that returned annotators, but no longer does, or other
+  /// such combinations, then you will either assert during the
+  /// subsequent call to [flushSemantics()] or you will have
+  /// out-of-date information in the semantics tree.
+  ///
+  /// If the geometry might have changed in any way, then again, more
+  /// work needs to be done to update the semantics tree (to deal with
+  /// clips). You can pass the noGeometry argument to avoid this work
+  /// in the case where only the labels or flags changed. If you pass
+  /// 'noGeometry: true' when the geometry did change, the semantic
+  /// tree will be out of date.
+  void markNeedsSemanticsUpdate({ bool onlyChanges: false, bool noGeometry: false }) {
+    assert(!_debugDoingSemantics);
+    if (!attached || (_needsSemanticsUpdate && onlyChanges &&
+                      (noGeometry || _needsSemanticsUpdateMarkedDescendants))) {
+      return;
+    }
+    if (!noGeometry && (_semantics == null || (_semantics.hasChildren && _semantics.wasAffectedByClip))) {
+      // Since the geometry might have changed, we need to make sure to reapply any clips.
+      visitChildrenForSemantics((RenderObject child) {
+        if (!(child._needsSemanticsUpdate && child._needsSemanticsUpdateMarkedDescendants))
+          child.markNeedsSemanticsUpdate(onlyChanges: true);
+      });
+      _needsSemanticsUpdateMarkedDescendants = true;
+    }
+    if (!onlyChanges) {
+      // We might be removing the node we've been using, or adding a
+      // new one and moving our parents' children down to us. Walk the
+      // tree up to our nearest ancestor with a SemanticsNode, and
+      // mark it dirty. Since removing it would affect its parent as
+      // well, mark that dirty too. And since removing it from _that_
+      // parent might cause that parent to only have one child, we
+      // need to tell _its_ parent as well. And if that parent was
+      // already dirty, then maybe it's got some other child that's
+      // going to be removed, in which case we should tell _it's_
+      // parent too. (This step recursively.)
+      RenderObject node = this;
+      node._needsSemanticsUpdate = true;
+      node._semantics?.reset();
+      int foundSemanticsNodes = 0;
+      do {
+        if (node._semantics != null)
+          foundSemanticsNodes += 1;
+        if (node.parent is! RenderObject)
+          break;
+        node = node.parent;
+        node._needsSemanticsUpdate = true;
+        node._semantics?.reset();
+      } while (node._needsSemanticsUpdate || node._semantics == null || foundSemanticsNodes < 3);
```

Comment:
We just set _needsSemanticsUpdate to true on line 1531.  How does this loop ever terminate?

Also, 3.